usim/py/events.py

```diff
@@ -0,0 +1,642 @@
+        return zip(self, self.values())
+
+    def todict(self):
+        return dict(self.items())
+
+
+class Condition(Event[ConditionValue]):
+    """
+    Event that is triggered when ``evaluate(events, num_triggered)`` is True
+
+    On any change of events, ``evaluate`` is called with two arguments:
+    ``events`` is a tuple of the events passed to the condition,
+    and ``num_triggered`` is the number of events that have triggered so far.
+    This can be used to quickly test how many events have triggered:
+
+    .. code:: python3
+
+        def most(events, count):
+            "Test that at least 50% of events have triggered"
+            return count * 2 >= len(events)
+
+        majority = Condition(most, events)
+
+    The value of a :py:class:`~.Condition` is a :py:class:`~.ConditionValue`
+    of **all** the events used in the condition. Note that events of nested
+    conditions are flattened. If *any* event fails, the :py:class:`~.Condition`
+    fails with the same reason.
+
+    .. note::
+
+        The condition is only evaluated on instantiation and when child events
+        trigger. This means that ``evaluate`` should not depend on external,
+        mutable objects.
+
+    .. hint::
+
+        **Migrating to μSim**
+
+        Since conditions, such as a :py:class:`~.Flag`, may change their value,
+        μSim does not support arbitrary comparisons. The common and well-defined
+        :py:meth:`~.all_of` and :py:meth:`~.any_of` correspond to the ``&`` and ``|``
+        operators.
+
+        .. code:: python3
+
+            all_events = flag1 & flag2 & flag3
+            any_events = flag1 | flag2 | flag3
+    """
+    __slots__ = '_evaluate', '_events', '_processed'
+
+    def __init__(self, env, evaluate, events: Iterable[Event]):
+        super().__init__(env)
+        self._evaluate = evaluate
+        self._events = tuple(events)
+        if any(event.env != env for event in self._events):
+            raise ValueError('Events from multiple environments cannot be mixed')
+        env.schedule(self._check_events(), delay=0)
+
+    async def _check_events(self):
+        observed = [event for event in self._events if event.__usimpy_flag__]
+        unobserved = [event for event in self._events if not event.__usimpy_flag__]
+        for event in observed:
+            if not event.ok:
+                # TODO: taken from simpy - this might swallow exceptions
+                #       if event is awaited but self is not awaited
+                event.defused = True
+                self.fail(event.value)
+                return
+        while unobserved and not self._evaluate(self._events, len(observed)):
+            await AnyFlag(*(event.__usimpy_flag__ for event in unobserved))
+            for event in unobserved[:]:
+                if not event.__usimpy_flag__:
+                    continue
+                elif event.ok:
+                    unobserved.remove(event)
+                    observed.append(event)
```

Comment:
From this point on you could think about only having a count for `observed` events as you are not accessing the events again. So you could introduce a local variable, initialise before while with `len(observed)` and then only do the increment.